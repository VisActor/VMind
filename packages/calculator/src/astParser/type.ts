import { Select } from 'node-sql-parser';
import { Query, TableData } from '../types';

export type SQLAst = Select;
export type ASTParserPipe = (query: Partial<Query>, context: ASTParserContext) => Partial<Query>;

export type ASTParserContext = {
  ast: SQLAst;
  dataSource: TableData;
  fieldInfo: SimpleFieldInfo[];
  replaceMap: Map<string, string>;
};

export type DataQueryResponse = {
  THOUGHT?: string;
  sql: string;
  fieldInfo: { fieldName: string; description?: string }[];
};

export type SimpleFieldInfo = {
  fieldName: string;
  description?: string; //additional description of the field. This will help the model have a more comprehensive understanding of this field, improving the quality of chart generation.
  type: DataType;
  role: ROLE;
};

export enum DataType {
  INT = 'int',
  STRING = 'string',
  FLOAT = 'float',
  DATE = 'date'
}

export enum ROLE {
  DIMENSION = 'dimension',
  MEASURE = 'measure'
}
