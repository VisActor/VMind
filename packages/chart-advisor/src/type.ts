//屏幕尺寸：小、中、大
export enum ScreenSize {
  LARGE = 0,
  MEDIUM = 1,
  SMALL = 2
}

//用户目的：对比、趋势、分布、排名、占比、组成、StoryTelling
export enum UserPurpose {
  NONE = 0, //未指定目的
  COMPARISON = 1,
  TREND = 2,
  DISTRIBUTION = 3,
  RANK = 4,
  PROPORTION = 5,
  COMPOSITION = 6,
  STORYTELLING = 7
}

export interface DimensionField {
  uniqueId: number; //该字段的id
  type: DataTypeName; //该字段的类型
  isGeoField?: boolean;
}

export interface MeasureField {
  uniqueId: number;
}

//measure数据集
export interface MeasureDataset {
  uniqueID?: number;
  data: number[];
  min?: number;
  max?: number;
  mean?: number; //平均值
  standardDev?: number; //标准差
  coefficient?: number; //变异系数
  Q1?: number; //下四分位数
}

//dimension数据集
export interface DimensionDataset {
  uniqueID?: number;
  data: string[];
  dataType?: DataTypeName;
  dimensionName?: string; //字段名
  cardinal?: number; //基数(不同值的数量)
  ratio?: number; //基数除以数据条数
  isGeoField?: boolean; // 是否为地理字段
}

export interface AutoChartCell {
  x: UniqueId[];
  y: UniqueId[];
  row: UniqueId[]; //作为透视行的字段
  column: UniqueId[]; //作为透视列的字段
  color?: UniqueId[];
  size?: UniqueId[];
  angle?: UniqueId[];
  value?: UniqueId[];
  text?: UniqueId[];
  group?: UniqueId[];
  error?: boolean;
  errMsg?: string;
  // 维度展开的信息（笛卡尔积）
  cartesianInfo?: CartesianInfo;
  // 指标展开的信息（指标平坦化）
  foldInfo?: FoldInfo;
}

export interface CartesianInfo {
  key: UniqueId;
  fieldList: UniqueId[];
}

export interface FoldInfo {
  key: UniqueId;
  value: UniqueId;
  foldMap: {
    [key: number]: string;
  };
}

export interface FieldTypeMap {
  [key: number]: DataTypeName;
}

export type DataItem = { [key: number]: string };

export type Dataset = DataItem[];

export type Datasets = DataItem[][][][] | Dataset;

export type UniqueId = number | string;

export type DataTypeName = 'number' | 'string' | 'date';

/**
 * vqs 接口中 visData 中的 aliasMap
 * 做字段名字映射
 */
export type AliasMap = {
  [key: number]: string;
};

/**
 * 图表类型枚举
 */
export enum ChartType {
  /** 表格 */
  TABLE = 'table',
  /** 明细表 */
  RAW_TABLE = 'raw_table',
  /** 透视表 */
  PIVOT_TABLE = 'pivot_table',

  /** 柱状图 */
  COLUMN = 'column',
  /** 百分比柱状图 */
  COLUMN_PERCENT = 'column_percent',
  /** 并列柱状图 */
  COLUMN_PARALLEL = 'column_parallel',
  /** 条形图 */
  BAR = 'bar',
  /** 百分比条形图 */
  BAR_PERCENT = 'bar_percent',
  /** 并列条形图 */
  BAR_PARALLEL = 'bar_parallel',

  /** 折线图 */
  LINE = 'line',
  /** 面积图 */
  AREA = 'area',
  /** 百分比面积图 */
  AREA_PERCENT = 'area_percent',

  /** 饼图 */
  PIE = 'pie',
  /** 环形饼图 */
  ANNULAR = 'annular',
  /** 南丁格尔玫瑰图 */
  ROSE = 'rose',

  /** 散点图 */
  SCATTER = 'scatter',
  /** 圆视图 */
  CIRCLE_VIEWS = 'circle_views',

  /** 双轴图 */
  DUAL_AXIS = 'double_axis',
  /** 双向条形图 */
  BILATERAL = 'bilateral',
  /** 组合图 */
  COMBINATION = 'combination',

  /** 填充地图 */
  MAP = 'map',
  /** 标记地图 */
  SCATTER_MAP = 'scatter_map',

  /** 指标卡 */
  MEASURE_CARD = 'measure_card',
  /** 对比指标卡 */
  COMPARATIVE_MEASURE_CARD = 'comparative_measure_card',
  /** 词云 */
  WORD_CLOUD = 'word_cloud',
  /** 直方图 */
  HISTOGRAM = 'histogram',
  /** 漏斗图 */
  FUNNEL = 'funnel',
  /** 雷达图 */
  RADAR = 'radar',
  /** 桑基图 */
  SANKEY = 'sankey',

  /** 扩展自定义类型 */
  EXTEND = 'extend',
  /** 单值百分比环形图 */
  PROGRESS = 'progress'
}

export interface ScoreResult {
  chartType: ChartType;
  originScore: number;
  fullMark: number;
  score: number;
  scoreDetails: Array<{ name: string; score: number }>;
  cell?: AutoChartCell | AutoChartCell[];
  dataset?: Datasets;
  error?: any;
}

export interface PivotTree {
  field: UniqueId;
  values: {
    value: string;
    child: PivotTree | null;
  }[];
}

export interface AdviseResult {
  chartType: ChartType; //vizData中的chartType
  scores: ScoreResult[];
  error?: any;
}

interface ScorerParams {
  inputDataSet: Dataset;
  dimList: DimensionDataset[];
  measureList: MeasureDataset[];
  aliasMap?: AliasMap;
  maxRowNum?: number;
  maxColNum?: number;
  purpose?: UserPurpose;
  screen?: ScreenSize;
}

export type Scorer = (params: ScorerParams) => Array<() => ScoreResult>;

export interface AdviserParams {
  originDataset: Dataset;
  dimensionList: DimensionField[];
  measureList: MeasureField[];
  aliasMap?: AliasMap;
  maxPivotRow?: number;
  maxPivotColumn?: number;
  purpose?: UserPurpose;
  screen?: ScreenSize;
  scorer?: Scorer;
}
