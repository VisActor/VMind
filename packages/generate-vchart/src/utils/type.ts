import dayjs from 'dayjs';

export function isInteger(value: any) {
  return typeof value === 'number' && Number.isInteger(value);
}

const dateFormats = [
  {
    key: 'YYYY年MM月DD日',
    regex: /^(\d{4})年(\d{1,2})月(\d{1,2})日$/,
    format: (match: string, year: string, month: string, day: string) =>
      `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`
  },

  {
    key: 'MM月DD日YYYY年',
    regex: /^(\d{1,2})月(\d{1,2})日(\d{4})年$/,
    format: (match: string, month: string, day: string, year: string) =>
      `${year}-${month.padStart(2, '0')}-${year.padStart(2, '0')}`
  },

  {
    key: 'YYYY年MM月',
    regex: /^(\d{4})年(\d{1,2})月$/,
    format: (match: string, year: string, month: string) => `${year}-${month.padStart(2, '0')}`
  },

  { key: 'YYYY年', regex: /^(\d{4})年$/, format: (match: string, year: string) => `${year}` },

  { key: 'MM月', regex: /^(\d{1,2})月$/, format: (match: string, month: string) => `${month.padStart(2, '0')}` },
  {
    key: 'MM月DD日',
    regex: /^(\d{1,2})月(\d{1,2})日$/,
    format: (match: string, month: string, day: string) => `${month.padStart(2, '0')}-${day.padStart(2, '0')}`
  }
];
function isQuarterDate(date: string) {
  const patterns = [/^\d{4}-Q[1-4]$/, /^\d{4}年第[一二三四]季度$/, /^Q[1-4]$/, /^\d{4}年-Q[1-4]$/];
  return patterns.some(pattern => pattern.test(date));
}

export function validateDate(date: any) {
  const formaterDate = `${date}`.trim();
  //check if the string is a data string
  return (
    dayjs(formaterDate).isValid() ||
    dateFormats.find(v => new Date(formaterDate.replace(v.regex, v.format)).toString() !== 'Invalid Date') ||
    isQuarterDate(formaterDate)
  );
}
