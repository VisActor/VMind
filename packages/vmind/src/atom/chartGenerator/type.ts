import type { ITheme as ChartTheme } from '@visactor/vchart';
import type { Cell, ChartGeneratorCtx, ChartType } from '../../types';
import type { BaseOptions } from '../../types/atom';
import type { BasemapOption, DataTable } from '@visactor/generate-vchart';

export type VisualChannel = 'x' | 'y' | 'color' | 'angle' | 'radius' | 'size' | 'value' | 'source' | 'target' | 'time';

export type ChartKnowledge = {
  [chartType: string]: {
    index: number;
    visualChannels: VisualChannel[]; //Visual channel available in this chart type
    examples: ((showThoughts: boolean) => string)[]; //examples of the response
    knowledge?: string[]; //extra knowledge of this chart
  };
};

export interface GenerateChartCellContext extends ChartGeneratorCtx {
  chartTypeList: ChartType[];
}

export interface ChartAdvistorResult {
  chartType: ChartType;
  cell?: Cell[];
  dataset?: DataTable[];
  spec: any;
  chartSource: string;
  usage: any; //token usage of the LLM
  time?: { totalTime: number; frameArr: any[] };
}

export interface ChartGeneratorOptions extends BaseOptions {
  useChartAdvisor?: boolean;
  useChartRule?: boolean;
  /** supported chart list */
  chartTypeList?: ChartType[];
  /** un-supported chart list */
  unsupportChartTypeList?: ChartType[];
  animationDuration?: number;
  basemapOption?: BasemapOption;
  colorPalette?: string[];
  theme?: ChartTheme | string;
}

export interface SimpleVChartSpecMockContext {
  ctx: Partial<GenerateChartCellContext>;
  mockLLMContent: {
    CHART_TYPE: ChartType;
    FIELD_MAP: Cell;
    stackOrPercent?: 'stack' | 'percent';
    transpose?: boolean;
  };
}
