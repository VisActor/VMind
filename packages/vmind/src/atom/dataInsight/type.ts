import type { Cell, ChartType } from '../../types/chart';
import type { DataCell, FieldInfo, DataTable, DataItem } from '../../types/base';
import type { AbnormalTrendOptions } from './algorithms/abnormalTrend';
import type { PearsonOptions } from './algorithms/correlation/pearson';
import type { OverallTrendingOptions } from './algorithms/overallTrending';
import type { SpearmanOptions } from './algorithms/correlation/spearman';
import type { ExtremeValueOptions } from './algorithms/extremeValue';
import type { MajorityValueOptions } from './algorithms/majorityValue';
import type { LOFOptions } from './algorithms/outlier/lof';
import type { DbscanOptions } from './algorithms/outlier/dbscan';
import type { VolatilityOptions } from './algorithms/volatility';
import type { StatisticsOptions } from './algorithms/outlier/statistics';
import type { DifferenceOptions } from './algorithms/outlier/difference';
import type { PageHinkleyOptions } from './algorithms/drift';

export enum AlgorithmType {
  OverallTrending = 'overallTrend',
  AbnormalTrend = 'abnormalTrend',
  PearsonCorrelation = 'pearsonCorrelation',
  SpearmanCorrelation = 'spearmanCorrelation',
  ExtremeValue = 'extremeValue',
  MajorityValue = 'majorityValue',
  StatisticsAbnormal = 'statisticsAbnormal',
  DbscanOutlier = 'dbscanOutlier',
  LOFOutlier = 'lofOutlier',
  TurningPoint = 'turningPoint',
  PageHinkley = 'pageHinkley',
  DifferenceOutlier = 'differenceOutlier',
  Volatility = 'volatility'
}

export interface AlgorithmOptions {
  [AlgorithmType.OverallTrending]?: OverallTrendingOptions;
  [AlgorithmType.AbnormalTrend]?: AbnormalTrendOptions;
  [AlgorithmType.PearsonCorrelation]?: PearsonOptions;
  [AlgorithmType.SpearmanCorrelation]: SpearmanOptions;
  [AlgorithmType.ExtremeValue]?: ExtremeValueOptions;
  [AlgorithmType.MajorityValue]?: MajorityValueOptions;
  [AlgorithmType.StatisticsAbnormal]?: StatisticsOptions;
  [AlgorithmType.LOFOutlier]?: LOFOptions;
  [AlgorithmType.DifferenceOutlier]?: DifferenceOptions;
  [AlgorithmType.TurningPoint]?: {};
  [AlgorithmType.PageHinkley]?: PageHinkleyOptions;
  [AlgorithmType.DbscanOutlier]: DbscanOptions;
  [AlgorithmType.Volatility]: VolatilityOptions;
}
export enum InsightType {
  Outlier = 'outlier',
  PairOutlier = 'pair_outlier',
  ExtremeValue = 'extreme_value',
  MajorityValue = 'majority_value',
  TurningPoint = 'turning_point',
  OverallTrend = 'overall_trend',
  AbnormalTrend = 'abnormal_trend',
  AbnormalBand = 'abnormal_band',
  Correlation = 'correlation',
  Volatility = 'volatility'
}

export interface InsightTextContent {
  value: DataCell;
  fieldName: string;
  color?: string;
  icon?: 'ratio' | 'ascendTrend' | 'descendTrend' | string;
}

export interface Insight {
  name: string; // algorithm name
  type: InsightType; // Insight type
  data: {
    // Abnormal data points
    index: number; // The index of the data point in the dataset
    dataItem: DataItem; // Data item
  }[];
  fieldId?: string; //abnormal field id
  seriesName?: DataCell | DataCell[]; //series name
  textContent?: {
    // The text explanation of the insight, generated by LLM, can be used for chart annotation.
    content: string; // Text content
    variables?: Record<string, InsightTextContent>;
  }; // The text array is composed of text fragments, and all elements in the array are combined to form a complete text.
  value?: number | string; // The specific value of the insight
  significant: number; // The prominence of insights, used for sorting
  info?: { [key: string]: any }; //additional information about this insight
}

export type InsightAlgorithm = {
  name: string; //specific algorithm name of this insight algorithm
  chartType?: ChartType[]; //supported chart type of this insight algorithm
  forceChartType?: ChartType[]; //force chart type of this insight algorithm
  /** support stack chart or not */
  supportStack?: boolean;
  /** support percent chart or not */
  supportPercent?: boolean;
  insightType: InsightType;
  canRun?: (context: DataInsightExtractContext) => boolean;
  algorithmFunction: (context: DataInsightExtractContext, options: any) => Insight[];
};

export type DataProcessOutput = {
  chartType: ChartType;
  cell: DataCell;
  fieldInfo: FieldInfo[];
  dataTable: DataTable;
  seriesDataMap: Record<string | number, { index: number; dataItem: DataItem }[]>;
};

export type ExtractInsightOutput = {
  insights: Insight[];
};

export type GenerateTextOutput = {
  insights: Insight[];
  usage: any;
};

export interface ChartDataItem {
  index: number;
  dataItem: DataItem;
}

export type DimValueDataMap = Record<string | number, ChartDataItem[]>;

export interface DataInsightExtractContext {
  dataset: DataTable;
  fieldInfo: FieldInfo[];
  dimensionDataMap: DimValueDataMap;
  dimensionValues: DataCell[];
  seriesDataMap: DimValueDataMap;
  chartType: ChartType;
  cell: Cell;
  spec: any;
  insights?: Insight[];
}
