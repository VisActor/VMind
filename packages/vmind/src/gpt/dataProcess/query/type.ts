import { Select } from 'node-sql-parser';
import { Query } from '../../../calculator';
import { DataItem, SimpleFieldInfo } from '../../../typings';

export type SQLAst = Select;
export type ASTParserPipe = (query: Partial<Query>, context: ASTParserContext) => Partial<Query>;

export type ASTParserContext = {
  ast: SQLAst;
  dataSource: DataItem[];
  fieldInfo: SimpleFieldInfo[];
  replaceMap: Map<string, string>;
};

export type DataQueryResponse = {
  THOUGHT?: string;
  sql: string;
  fieldInfo: { fieldName: string; description?: string }[];
};
