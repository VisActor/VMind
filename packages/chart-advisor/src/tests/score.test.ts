import { scorer } from '../score';
import { ChartType, UserPurpose, ScreenSize } from '../type';

describe('scorer', () => {
  it('should return COLUMN chart type for basic bar scenario', () => {
    const dimList = [
      {
        uniqueID: 1,
        data: ['A', 'B', 'C'],
        dataType: 'string' as const,
        cardinal: 3
      }
    ];
    const measureList = [
      {
        uniqueID: 2,
        data: [10, 20, 30],
        min: 10,
        max: 30,
        mean: 20,
        standardDev: 10,
        coefficient: 0.5,
        Q1: 10
      }
    ];
    const inputDataSet = [
      { 1: 'A', 2: '10' },
      { 1: 'B', 2: '20' },
      { 1: 'C', 2: '30' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[0]();
    expect(result.chartType).toBe(ChartType.COLUMN);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return COLUMN_PERCENT chart type', () => {
    const dimList = [
      { uniqueID: 1, data: ['A', 'B'], dataType: 'string' as const, cardinal: 2 },
      { uniqueID: 2, data: ['X', 'Y'], dataType: 'string' as const, cardinal: 2 }
    ];
    const measureList = [
      { uniqueID: 3, data: [10, 20, 30, 40], min: 10, max: 40, mean: 25, standardDev: 12, coefficient: 0.48, Q1: 10 }
    ];
    const inputDataSet = [
      { 1: 'A', 2: 'X', 3: '10' },
      { 1: 'A', 2: 'Y', 3: '20' },
      { 1: 'B', 2: 'X', 3: '30' },
      { 1: 'B', 2: 'Y', 3: '40' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[1]();
    expect(result.chartType).toBe(ChartType.COLUMN_PERCENT);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return COLUMN_PARALLEL chart type', () => {
    const dimList = [{ uniqueID: 1, data: ['A', 'B'], dataType: 'string' as const, cardinal: 2 }];
    const measureList = [
      { uniqueID: 2, data: [10, 20], min: 10, max: 20, mean: 15, standardDev: 5, coefficient: 0.33, Q1: 10 },
      { uniqueID: 3, data: [15, 25], min: 15, max: 25, mean: 20, standardDev: 5, coefficient: 0.25, Q1: 15 }
    ];
    const inputDataSet = [
      { 1: 'A', 2: '10', 3: '15' },
      { 1: 'B', 2: '20', 3: '25' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[2]();
    expect(result.chartType).toBe(ChartType.COLUMN_PARALLEL);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return SCATTER chart type', () => {
    const dimList = [
      {
        uniqueID: 1,
        data: [
          'A',
          'B',
          'C',
          'D',
          'E',
          'F',
          'G',
          'H',
          'I',
          'J',
          'K',
          'L',
          'M',
          'N',
          'O',
          'P',
          'Q',
          'R',
          'S',
          'T',
          'U',
          'V',
          'W',
          'X',
          'Y',
          'Z',
          'AA',
          'AB',
          'AC',
          'AD'
        ],
        dataType: 'string' as const,
        cardinal: 30
      }
    ];
    const measureList = [
      {
        uniqueID: 2,
        data: Array(30)
          .fill(1)
          .map((_, i) => i + 1),
        min: 1,
        max: 30,
        mean: 15.5,
        standardDev: 8.7,
        coefficient: 0.56,
        Q1: 8
      },
      {
        uniqueID: 3,
        data: Array(30)
          .fill(2)
          .map((_, i) => i + 2),
        min: 2,
        max: 31,
        mean: 16.5,
        standardDev: 8.7,
        coefficient: 0.53,
        Q1: 9
      }
    ];
    const inputDataSet = Array(30)
      .fill(0)
      .map((_, i) => ({ 1: String.fromCharCode(65 + (i % 26)), 2: String(i + 1), 3: String(i + 2) }));
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[3]();
    expect(result.chartType).toBe(ChartType.SCATTER);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return LINE chart type', () => {
    const dimList = [
      { uniqueID: 1, data: ['2020-01-01', '2020-01-02', '2020-01-03'], dataType: 'date' as const, cardinal: 3 }
    ];
    const measureList = [
      { uniqueID: 2, data: [10, 20, 30], min: 10, max: 30, mean: 20, standardDev: 10, coefficient: 0.5, Q1: 10 }
    ];
    const inputDataSet = [
      { 1: '2020-01-01', 2: '10' },
      { 1: '2020-01-02', 2: '20' },
      { 1: '2020-01-03', 2: '30' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[4]();
    expect(result.chartType).toBe(ChartType.LINE);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return PIE chart type for pie scenario', () => {
    const dimList = [];
    const measureList = [
      { uniqueID: 2, data: [10, 20, 30], min: 10, max: 30, mean: 20, standardDev: 10, coefficient: 0.5, Q1: 10 },
      { uniqueID: 3, data: [5, 15, 25], min: 5, max: 25, mean: 15, standardDev: 10, coefficient: 0.66, Q1: 5 },
      { uniqueID: 4, data: [8, 18, 28], min: 8, max: 28, mean: 18, standardDev: 10, coefficient: 0.55, Q1: 8 }
    ];
    const inputDataSet = [
      { 2: '10', 3: '5', 4: '8' },
      { 2: '20', 3: '15', 4: '18' },
      { 2: '30', 3: '25', 4: '28' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const pieResult = resultFns[5]();
    expect(pieResult.chartType).toBe(ChartType.PIE);
    expect(pieResult.score).toBeGreaterThanOrEqual(0);
    expect(pieResult.fullMark).toBeGreaterThan(0);
  });

  it('should return RADAR chart type', () => {
    const dimList = [{ uniqueID: 1, data: ['A', 'B'], dataType: 'string' as const, cardinal: 2 }];
    const measureList = [
      { uniqueID: 2, data: [10, 20], min: 10, max: 20, mean: 15, standardDev: 5, coefficient: 0.33, Q1: 10 }
    ];
    const inputDataSet = [
      { 1: 'A', 2: '10' },
      { 1: 'B', 2: '20' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList, purpose: UserPurpose.DISTRIBUTION });
    const result = resultFns[6]();
    expect(result.chartType).toBe(ChartType.RADAR);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return WORD_CLOUD chart type', () => {
    const dimList = [
      {
        uniqueID: 1,
        data: Array(20)
          .fill(0)
          .map((_, i) => `word${i}`),
        dataType: 'string' as const,
        cardinal: 20
      }
    ];
    const measureList = [];
    const inputDataSet = Array(20)
      .fill(0)
      .map((_, i) => ({ 1: `word${i}` }));
    const resultFns = scorer({ inputDataSet, dimList, measureList, purpose: UserPurpose.STORYTELLING });
    const result = resultFns[7]();
    expect(result.chartType).toBe(ChartType.WORD_CLOUD);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return FUNNEL chart type', () => {
    const dimList = [{ uniqueID: 1, data: ['A'], dataType: 'string' as const, cardinal: 1 }];
    const measureList = [
      { uniqueID: 2, data: [10], min: 10, max: 10, mean: 10, standardDev: 0, coefficient: 0, Q1: 10 }
    ];
    const inputDataSet = [{ 1: 'A', 2: '10' }];
    const resultFns = scorer({ inputDataSet, dimList, measureList, purpose: UserPurpose.TREND });
    const result = resultFns[8]();
    expect(result.chartType).toBe(ChartType.FUNNEL);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  it('should return DUAL_AXIS chart type', () => {
    const dimList = [{ uniqueID: 1, data: ['A', 'B'], dataType: 'string' as const, cardinal: 2 }];
    const measureList = [
      { uniqueID: 2, data: [1, 200], min: 1, max: 200, mean: 100.5, standardDev: 140.007, coefficient: 1.393, Q1: 1 },
      { uniqueID: 3, data: [300, 400], min: 300, max: 400, mean: 350, standardDev: 70.71, coefficient: 0.202, Q1: 300 }
    ];
    const inputDataSet = [
      { 1: 'A', 2: '1', 3: '300' },
      { 1: 'B', 2: '200', 3: '400' }
    ];
    const resultFns = scorer({ inputDataSet, dimList, measureList });
    const result = resultFns[9]();
    expect(result.chartType).toBe(ChartType.DUAL_AXIS);
    expect(result.score).toBeGreaterThanOrEqual(0);
    expect(result.fullMark).toBeGreaterThan(0);
  });

  // 边界与异常场景
  describe('异常与边界场景', () => {
    it('should handle empty inputDataSet', () => {
      const dimList = [{ uniqueID: 1, data: [], dataType: 'string' as const, cardinal: 0 }];
      const measureList = [{ uniqueID: 2, data: [], min: 0, max: 0, mean: 0, standardDev: 0, coefficient: 0, Q1: 0 }];
      const inputDataSet = [];
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      resultFns.forEach(fn => {
        const result = fn();
        expect(result.score).toBe(0);
        expect(result.fullMark).toBe(0);
        expect(result.chartType).toBeDefined();
      });
    });

    it('should handle empty dimList', () => {
      const dimList = [];
      const measureList = [
        { uniqueID: 2, data: [10, 20], min: 10, max: 20, mean: 15, standardDev: 5, coefficient: 0.33, Q1: 10 }
      ];
      const inputDataSet = [{ 2: '10' }, { 2: '20' }];
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      resultFns.forEach(fn => {
        const result = fn();
        expect(result.chartType).toBeDefined();
      });
    });

    it('should handle empty measureList', () => {
      const dimList = [{ uniqueID: 1, data: ['A', 'B'], dataType: 'string' as const, cardinal: 2 }];
      const measureList = [];
      const inputDataSet = [{ 1: 'A' }, { 1: 'B' }];
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      resultFns.forEach(fn => {
        const result = fn();
        expect(result.chartType).toBeDefined();
      });
    });

    it('should handle mismatched data types', () => {
      const dimList = [{ uniqueID: 1, data: ['A', 'B'], dataType: 'date' as const, cardinal: 2 }];
      const measureList = [
        { uniqueID: 2, data: [10, 20], min: 10, max: 20, mean: 15, standardDev: 5, coefficient: 0.33, Q1: 10 }
      ];
      const inputDataSet = [
        { 1: 'A', 2: '10' },
        { 1: 'B', 2: '20' }
      ];
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      resultFns.forEach(fn => {
        const result = fn();
        expect(result.chartType).toBeDefined();
      });
    });

    it('should handle large data size', () => {
      const dimList = [
        {
          uniqueID: 1,
          data: Array(101)
            .fill(0)
            .map((_, i) => `A${i}`),
          dataType: 'string' as const,
          cardinal: 101
        }
      ];
      const measureList = [
        { uniqueID: 2, data: Array(101).fill(1), min: 1, max: 1, mean: 1, standardDev: 0, coefficient: 0, Q1: 1 }
      ];
      const inputDataSet = Array(101)
        .fill(0)
        .map((_, i) => ({ 1: `A${i}`, 2: '1' }));
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      resultFns.forEach(fn => {
        const result = fn();
        expect(result.chartType).toBeDefined();
      });
    });
  });

  // 结构断言与快照
  describe('结构断言与快照', () => {
    it('COLUMN chartType result structure', () => {
      const dimList = [{ uniqueID: 1, data: ['A', 'B', 'C'], dataType: 'string' as const, cardinal: 3 }];
      const measureList = [
        { uniqueID: 2, data: [10, 20, 30], min: 10, max: 30, mean: 20, standardDev: 10, coefficient: 0.5, Q1: 10 }
      ];
      const inputDataSet = [
        { 1: 'A', 2: '10' },
        { 1: 'B', 2: '20' },
        { 1: 'C', 2: '30' }
      ];
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      const result = resultFns[0]();
      expect(result.cell).toBeDefined();
      expect(result.dataset).toBeDefined();
      expect(result.scoreDetails).toBeDefined();
      expect(result).toMatchSnapshot();
    });
  });

  // 参数化测试示例
  describe.each([
    ['COLUMN', 0, ChartType.COLUMN],
    ['COLUMN_PERCENT', 1, ChartType.COLUMN_PERCENT],
    ['COLUMN_PARALLEL', 2, ChartType.COLUMN_PARALLEL],
    ['SCATTER', 3, ChartType.SCATTER],
    ['LINE', 4, ChartType.LINE],
    ['PIE', 5, ChartType.PIE],
    ['RADAR', 6, ChartType.RADAR],
    ['WORD_CLOUD', 7, ChartType.WORD_CLOUD],
    ['FUNNEL', 8, ChartType.FUNNEL],
    ['DUAL_AXIS', 9, ChartType.DUAL_AXIS]
  ])('主流类型参数化断言: %s', (desc, idx, expectedType) => {
    it(`should return correct chartType for ${desc}`, () => {
      // 这里用最小 mock 数据，实际可根据 idx 定制
      // 以 COLUMN 为例
      let dimList, measureList, inputDataSet;
      if (expectedType === ChartType.COLUMN) {
        dimList = [{ uniqueID: 1, data: ['A', 'B', 'C'], dataType: 'string' as const, cardinal: 3 }];
        measureList = [
          { uniqueID: 2, data: [10, 20, 30], min: 10, max: 30, mean: 20, standardDev: 10, coefficient: 0.5, Q1: 10 }
        ];
        inputDataSet = [
          { 1: 'A', 2: '10' },
          { 1: 'B', 2: '20' },
          { 1: 'C', 2: '30' }
        ];
      } else {
        // 其它类型可用已有测试数据或简化 mock
        dimList = [];
        measureList = [
          { uniqueID: 2, data: [10, 20, 30], min: 10, max: 30, mean: 20, standardDev: 10, coefficient: 0.5, Q1: 10 }
        ];
        inputDataSet = [{ 2: '10' }, { 2: '20' }, { 2: '30' }];
      }
      const resultFns = scorer({ inputDataSet, dimList, measureList });
      const result = resultFns[idx]();
      expect(result.chartType).toBe(expectedType);
    });
  });
});
