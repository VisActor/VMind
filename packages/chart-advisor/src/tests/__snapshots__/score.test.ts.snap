// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scorer 结构断言与快照 COLUMN chartType result structure 1`] = `
{
  "cell": {
    "angle": [],
    "color": [],
    "column": [],
    "row": [],
    "size": [],
    "x": [
      1,
    ],
    "y": [
      2,
    ],
  },
  "chartType": "column",
  "dataset": [
    {
      "1": "A",
      "2": "10",
    },
    {
      "1": "B",
      "2": "20",
    },
    {
      "1": "C",
      "2": "30",
    },
  ],
  "fullMark": 7,
  "originScore": 6,
  "score": 0.8571428571428571,
  "scoreDetails": {
    "rule1": 1,
    "rule2": 3,
    "rule3": 1,
    "rule4": 1,
  },
}
`;
